File: sdk/python/src/spacerouter/identity.py
"""Node identity keypair management for the SpaceRouter SDK.

Generates and persists a secp256k1 keypair used for signing authenticated
API requests.  Default storage: ``~/.spacerouter/identity.key``.
"""

import os
import time
from pathlib import Path

from eth_account import Account
from eth_account.messages import encode_defunct
from web3 import Web3

_w3 = Web3()

DEFAULT_IDENTITY_PATH = str(Path.home() / ".spacerouter" / "identity.key")


def load_or_create_identity(key_path: str = DEFAULT_IDENTITY_PATH) -> tuple[str, str]:
    """Load or generate a secp256k1 identity keypair.

    Returns ``(private_key_hex, address)``.
    """
    if os.path.isfile(key_path):
        with open(key_path) as f:
            private_key = f.read().strip()
        account = Account.from_key(private_key)
        return private_key, account.address.lower()

    account = Account.create()
    private_key = account.key.hex()

    os.makedirs(os.path.dirname(key_path) or ".", exist_ok=True)
    with open(key_path, "w") as f:
        f.write(private_key + "\n")
    os.chmod(key_path, 0o600)

    return private_key, account.address.lower()


def get_address(private_key: str) -> str:
    """Derive the checksummed-lower Ethereum address from a private key."""
    return Account.from_key(private_key).address.lower()


def sign_request(private_key: str, action: str, target: str) -> tuple[str, int]:
    """Sign a Space Router API request.

    Returns ``(signature_hex, timestamp)``.
    """
    timestamp = int(time.time())
    message_text = f"space-router:{action}:{target}:{timestamp}"
    message = encode_defunct(text=message_text)
    signed = _w3.eth.account.sign_message(message, private_key=private_key)
    return signed.signature.hex(), timestamp


def create_vouching_signature(
    private_key: str, staking_address: str, collection_address: str,
) -> tuple[str, int]:
    """Sign a vouching message: identity wallet vouches for staking + collection wallets.

    Message format: ``space-router:vouch:{staking_address}:{collection_address}:{timestamp}``

    Returns ``(signature_hex, timestamp)``.
    """
    timestamp = int(time.time())
    message_text = f"space-router:vouch:{staking_address.lower()}:{collection_address.lower()}:{timestamp}"
    message = encode_defunct(text=message_text)
    signed = _w3.eth.account.sign_message(message, private_key=private_key)
    return signed.signature.hex(), timestamp

File: sdk/js/src/identity.ts
/**
 * Node identity keypair management for the SpaceRouter JS SDK.
 *
 * Generates and persists a secp256k1 keypair used for signing authenticated
 * API requests.  Default storage: `~/.spacerouter/identity.key`.
 */

import {
  privateKeyToAccount,
  generatePrivateKey,
} from "viem/accounts";
import {
  readFileSync,
  writeFileSync,
  mkdirSync,
  existsSync,
  chmodSync,
} from "node:fs";
import { dirname, join } from "node:path";
import { homedir } from "node:os";

const DEFAULT_IDENTITY_PATH = join(
  homedir(),
  ".spacerouter",
  "identity.key",
);

/**
 * Load or generate a secp256k1 identity keypair.
 *
 * Returns `{ privateKey, address }`.
 */
export function loadOrCreateIdentity(
  keyPath: string = DEFAULT_IDENTITY_PATH,
): { privateKey: `0x${string}`; address: string } {
  if (existsSync(keyPath)) {
    const raw = readFileSync(keyPath, "utf-8").trim();
    const privateKey = (
      raw.startsWith("0x") ? raw : `0x${raw}`
    ) as `0x${string}`;
    const account = privateKeyToAccount(privateKey);
    return { privateKey, address: account.address.toLowerCase() };
  }

  const privateKey = generatePrivateKey();
  const account = privateKeyToAccount(privateKey);

  mkdirSync(dirname(keyPath), { recursive: true });
  writeFileSync(keyPath, privateKey + "\n", { mode: 0o600 });
  try {
    chmodSync(keyPath, 0o600);
  } catch {
    // chmod may fail on some platforms — best-effort
  }

  return { privateKey, address: account.address.toLowerCase() };
}

/** Derive the lowercase Ethereum address from a private key. */
export function getAddress(privateKey: `0x${string}`): string {
  return privateKeyToAccount(privateKey).address.toLowerCase();
}

/**
 * Sign a Space Router API request using EIP-191.
 *
 * Message format: `space-router:{action}:{target}:{timestamp}`
 *
 * Returns `{ signature, timestamp }`.
 */
export async function signRequest(
  privateKey: `0x${string}`,
  action: string,
  target: string,
): Promise<{ signature: string; timestamp: number }> {
  const timestamp = Math.floor(Date.now() / 1000);
  const message = `space-router:${action}:${target}:${timestamp}`;
  const account = privateKeyToAccount(privateKey);
  const signature = await account.signMessage({ message });
  return { signature, timestamp };
}

/**
 * Sign a vouching message: identity wallet vouches for staking wallet.
 *
 * Message format: `space-router:vouch:{stakingAddress}:{timestamp}`
 *
 * Returns `{ signature, timestamp }`.
 */
export async function createVouchingSignature(
  privateKey: `0x${string}`,
  stakingAddress: string,
): Promise<{ signature: string; timestamp: number }> {
  const timestamp = Math.floor(Date.now() / 1000);
  const message = `space-router:vouch:${stakingAddress.toLowerCase()}:${timestamp}`;
  const account = privateKeyToAccount(privateKey);
  const signature = await account.signMessage({ message });
  return { signature, timestamp };
}
